### DISCUSSION TITLE
Ensuring the integrity of firmware

### DISCUSSION THREAD

**Discussion by mYnDstrEAm** [fwupd/fwupd#4261] [General]
**Is your feature request related to a problem? Please describe.**
For packages in Debian's repositories, hashes and signatures are checked as described [here](https://unix.stackexchange.com/questions/285635/how-is-the-authenticity-of-debian-packages-guaranteed) (and questions linked there) to ensure the integrity of the software to at least some degree (under development).

I asked about it [here](https://unix.stackexchange.com/questions/689663/how-can-the-integrity-of-firmware-be-ensured-preinstalled-and-fwupd-updates) and it seems like fwupd ensures the integrity of firmware updates only by some measures that aim to ensure authenticity via LVFS (and I don't know how reliable that method is or how it works in detail).

I think there broadly would only be one way to reliably ensure integrity - these three points are necessary:
* the firmware being open source and [reproducible](https://reproducible-builds.org/) (so building from the same source code results in artifacts with always the same hashsums, ensuring that their code is unmodified and genuine)
* the reproducible hashsums of the files being public
* a reliable way to check the local firmware against these hashsums.

This issue is currently mainly about the third point. Does fwupd check the downloaded software against signed hashsums to ensure authenticity?

Could you add mechanisms that better ensure that the hashsums of the local downloaded files really match the firmware hashsum and that these firmware hashsums are globally the same? It could also be useful if firmware software providers could be required to ensure integrity to at least some level such as by being required to upload the firmware from multiple specified machines with files built on multiple, not just one for example, build-machines and signed by multiple specified keys with the firmware update for example becoming available after a one-day wait-time. 

This is a large and very important issue, so if it's phrased not sufficiently well, maybe a new issue should be created at some point.

**Describe the solution you'd like**
For example: an improvement of how hashsums are validated locally with fwupd and this issue getting moved elsewhere (like LVFS) afterwards if further improvements would depend on changes there.

**Describe alternatives you've considered**

**Additional context**
I'm aware that lots of firmware is not open source so the optimal solution can't be applied for all hardware but only some devices. As a side note it would be useful if `fwupdmgr security --force` also showed which firmware is open or closed source but that's not part of this issue. Functionality to check the integrity of already installed firmware could also be added.

**Comment by hughsie** [comment]
Are you asking about:

* download integrity, i.e. if the firmware files were downloaded correctly
* distributor integrity, i.e. all the firmware was distributed via the LVFS
* update integrity, i.e. the image was written to the onboard SPI device correctly
* system attestation. i.e. the image that we wrote last week is still the same image that's running now

Each is it's own distinct problem, with it's own solution.

**Reply by mYnDstrEAm** [reply]
Well the issue is very comprehensive so it's probably about all of these (that what I meant with it being a large issue). It's about ensuring integrity of firmware with all the requirements this has (the only thing I'd exclude would be vulnerabilities of compilers and things like that as these would be an issue above the level of "firmware integrity").

For more immediate work the issue is about download integrity and integrity of the files uploaded onto LVFS.

I think "system attestation" would be part of "Functionality to check the integrity of already installed firmware" which I'd mostly consider "additional context" insofar as this could be implemented afterwards: making sure that the firmware remains unchanged is probably also a very comprehensive issue. In the same way, "update integrity" could be implemented afterwards once the integrity of the files to write to the device is sufficiently ensured via hashsums and authentications.

**Reply by hughsie** [reply]
> download integrity

We've got SHA1 and SHA256 verifications for those.

> integrity of the files uploaded onto LVFS

That's JCAT, which is a thin wrapper around PKCS#7 and GPG -- by default we use either.

> making sure that the firmware remains unchanged

The best we have for system firmware is the PCR0, and for some peripheral devices we don't even have that.

**Reply by mYnDstrEAm** [reply]
> We've got SHA1 and SHA256 verifications for those.

To which signatures does the local fwupd check these against? Are they reliably always the same (and not vary in time or by region or by server)? It would be useful to have a little more detail like the linked explanation for how it works with package managers in Debian (such as a page that describes this).

>That's JCAT, which is a thin wrapper around PKCS#7 and GPG -- by default we use either.

I don't know what JCAT is and how all of this works. For example, how is it ensured (referring to the current suboptimal mechanisms) that there's not multiple signed firmwares and a malicious (due to modification) one is downloaded? I addressed things like that in the original post.  
For instance, a mechanism of _all firmware is signed with one GPG key each_ does not address that the one GPG key may have signed malicious software (wittingly or unwittingly or not done by the person publicly associated with the key) or that malicious software was also signed by some GPG keys and uploaded.

>The best we have for system firmware is the PCR0, and for some peripheral devices we don't even have that.

I guess it's best to set that aside for now here (referring to this issue and any subtasks) until integrity of newly installed firmware is ensured. So that would be a separate issue (ensuring firmware remains unaltered is probably also a very comprehensive issue).

**Reply by hughsie** [reply]
> To which signatures does the local fwupd check these against

This is checking the download against the remote manifest, e.g. /var/lib/fwupd/remotes.d/lvfs/metadata.xml.gz

> Are they reliably always the same (and not vary in time or by region or by server)?

Yes, there's a global CDN.

> how it works with package managers in Debian

It's completely orthogonal to your package manager, it doesn't interact with that in any way.

> I don't know what JCAT is and how all of this works.

This is required reading: https://github.com/hughsie/libjcat/blob/main/README.md

> does not address that the one GPG key may have signed malicious software

Sure, the JCAT signature is verifying that the file was uploaded by a trusted OEM and uploaded and distributed from the LVFS. It's not replacing the signing key from the vendor, it's completely detached.
